### ISSUE TITLE
GeoJSON with popups and supercluster

### ISSUE THREAD

**Issue by nistran** [thedirtyfew/dash-leaflet#41]
Hi
First of all I did the update to geoJSON and the improvement of performance is really nice :+1: . 

Now I stared having popups on my map like it is described in #37 . 

In fact it works like a charm, but I have some issues together with the supercluster. 
Now I get a popup on both, clicking on a cluster and a GPS tag. Unfortunately this leads into some problems. 
To fix this I have two ideas:
First: 
Is there a way to control the popups? So that I'm able to prevent them when clicking on a cluster.
Second: 
Is there a way to get the GPS points a cluster includes? In the supercluster documentation are methods like 
`getLeaves(clusterId, limit = 10, offset = 0)`
In fact I have the cluster_id from the click_feature callback. But I have currently no idea how to access the points which are included in the cluster. 
With this possibility I would be able to display all images which are inside the cluster when the popup is displayed. 

The more I think about this two ideas I highly appreciate the second idea ;)

I will be happy getting some ideas to solve that.

**Comment by emilhe** [comment]
Thanks! I am aware of this issue, and i am currently thinking about what the proper solution would be.

1) I am not sure if it is possible to control tooltips/popups for individual elements when the tooltip/popup is added to the `GeoJSON` component itself. This would be the preferred options, so if you come across any information, please let me know. 

2) If you don't attach the tooltip to the `GeoJSON` component itself, but rather at the individual elements are they are rendered, you can do pretty much anything you like (including skipping popups for clusters or rendering its properties related to its leaves). However, to do this, you would need the write your own `onEachFeature` function in JavaScript and bind it to the `GeoJSON` component. 

\emher

**Comment by nistran** [comment]
@emilhe 
Thanks for the fast relpy. The only other idea to control the popups would be the filter optrio

**Comment by nistran** [comment]
Sry I hit the wrong button 

Thanks for the fast answer. I think I'll give the JavaScript option a try. I think the method is overwritten like you described in you tutorials?

**Comment by emilhe** [comment]
Yes, it should be more or less straight forward. The structure of the javascript would be something like,

```
window.dash_props = Object.assign({}, window.dash_props, {
    module: {
        on_each_feature: function (feature, layer, context) {
            // Add popup.
            if (feature.properties.popup) {
                const el = layer.bindPopup(feature.properties.popup).openPopup()
                // Check if feature id is matching the id for which the popup should be open.
                if(context.props.hideout && feature.properties.id === context.props.hideout.open){
                    const {map} = context.props.leaflet;  // get the map object
                    map.addLayer(el);  // add layer (otherwise we can't open the popup)
                    el.openPopup();  // open the popup
                }
            }
            // Add tooltip.
            if (feature.properties.tooltip) {
                layer.bindTooltip(feature.properties.tooltip)
            }
        }
    }
});
```

and to bind the function to the GeoJSON component, you would simple provide the function path,

```
        dl.GeoJSON(... options=dict(onEachFeature="window.dash_props.module.on_each_feature"))
```

**Comment by nistran** [comment]
@emilhe thanks for providing that code snipped. I now found the the time to give that a try. So far so good i managed to bind the js function to on_each_feature and print some messages to the console.  But now I'm wondering how I manage to have an individual popup, which i can handle from there. As i understood I'm able to use the hideout property to pass variables to the js function. But how I do manage to open a Popup and fill it with information? For now when I'm trying to print the variable ```feature.properties.popup``` I receive undefined. I'm a c++ developer and a absolute newbie to java script. So the learning curve is currently quite high.

**Comment by nistran** [comment]
Ok I think I have understood how to manage the popups with js. So just ignore my last comment.

**Comment by emilhe** [comment]
What solution did you end up with?

**Comment by nistran** [comment]
Unfortunately I ended up with no nice solution, so I think for now I end up with the solution I had before the js function stuff.. The problem is, that I'm able to open a popup and control the content, but its is hard to combine this with the dash world. Positive is that I can skip popups on cluster clicks and get only popups on GPS Tags. 
Maybe I describe the workflow I had before starting with the js function bindings:
The user clicks on a GPS Tag and this opens an empty popup triggerd from Dash Leaflet. At the sams time (or nearly the same time) I receive a dash callback for the click event. The output is the div id of the empty popup, which I bind to the geoJson stuff.Then I need to prepare some date and generate some tokens and such stuff inside this callback. This is filled into a html.div tree which describes the popup. 
This works quite well with two drawbacks. First I receive popups for cluster clicks which I can't hanld in a nice way. Second there is a short delay between the empty popup is opened from leaflet and I'm ready with preparing the data to fill into that popup with the callback. So the user gets a empty popup or popup with old data for a short time. 
So why is the Js Function binding way not suitable for me? I'm not able to find a way, how to get the on demand prepared data into that popup, which is created inside the jd function. Dash callback outputs to all that stuff, which is generated inside this js function is not possible. 
So in short, I can't find a way how to process data on the click event in python and use this in the js function to fill the popup with that data. 
So this was I quiet long text. I hope where able to understand what I was trying to say.

**Comment by nistran** [comment]
UPDATE: 
I now decided to feed the tooltips with the picture urls when building the geojson structure. For now I'm quite happy with this solution. 
I think we can close this issue for now?

**Comment by emilhe** [comment]
Could you post a piece of example code? I would like to see how it turned out :)

**Comment by nistran** [comment]
Hi sry for the late repay
Basically it is almost the same code then your geojson examples. 
Additional I added the image as tooltip in pure html

```py
item['tooltip'] = f"""<b>Name:</b> {name}<br><b>Description:</b> {description}<br><img src="{url}"style="object-fit: scale-down;width:200px;">"""
```
and the following JavaScript module to remove the opacity of the tooltips

```js
window.dash_props = Object.assign({}, window.dash_props, {
    module: {
        on_each_feature: function(feature, layer, context) {
            if (!feature.properties) {
                return
            }
            if (feature.properties.popup) {
                layer.bindPopup(feature.properties.popup)
            }
            if (feature.properties.tooltip) {
                // here you can change all leaflet tooltip options
                layer.bindTooltip(feature.properties.tooltip, { opacity: 1.0 })
            }
        }
    }
});
```

**Comment by nistran** [comment]
@emilhe With the latest version 0.1.10 I now have exceptions with 
```
options=dict(onEachFeature="window.dash_props.module.on_each_feature")
```
I get the following error: 
```
Uncaught TypeError: options.onEachFeature is not a function
    at NewClass.addData (leaflet-src.js?e11e:8704)
    at NewClass.addData (leaflet-src.js?e11e:8684)
    at LeafletGeoJSON._render_clusters (LeafletGeoJSON.js?3d90:197)
    at NewClass.fire (leaflet-src.js?e11e:588)
    at NewClass._moveEnd (leaflet-src.js?e11e:4223)
    at NewClass.eval (leaflet-src.js?e11e:4675)
```
Do you have a solution for this?
Thanks in advance !

**Comment by emilhe** [comment]
Sorry, that's due to a (breaking) syntax change. Using `dash-extensions==0.0.32` and `dash-leaflet==0.1.10`, the new syntax would be,

```
from dash_extensions.javascript import Namespace
...
ns = Namespace("dash_props", "module")
options=dict(onEachFeature=ns("on_each_feature")
```

**Comment by nistran** [comment]
@emilhe Thanks for the fast response. Your suggestion fixed that issue for me.
Nerveless I think we can close that issue, for now.
